rustc_target: callconv: powerpc64: Use the ABI set in target options instead of guessing

Gelbpunkt opened the pull request:
*[View all comments](https://triagebot.infra.rust-lang.org/gh-comments/rust-lang/rust/pull/150468)*

All PowerPC64 targets except AIX explicitly set the ABI in the target options. We can therefore stop hardcoding the ABI to be used based on the target environment or OS, except for the AIX special case.

The fallback based on endianness is kept for the sake of compatibility with custom targets.

This makes it so that big endian targets not explicitly accounted for before (powerpc64-unknown-openbsd) and targets that don't use the expected default ABI (big-endian ELFv2 Glibc targets) use the correct ABI in the calling convention code.

The second commit is a tiny change to validate the `llvm_abiname` set on PowerPC64(LE) targets. See the commit messages for details.

CC @RalfJung who pointed out the missing `llvm_abiname` validation

chenyukang commented:
@rustbot reroll

taiki-e commented on compiler/rustc_target/src/spec/mod.rs:
```diff
@@ -3178,6 +3178,22 @@ impl Target {
+                );
+            }
+            // Little-endian PowerPC64 only supports the ELFv2 ABI
+            Arch::PowerPC64LE => {
```
`Arch::PowerPC64LE` is unused and `Arch` for Little-endian PowerPC64 is `Arch::PowerPC64`. (See https://github.com/rust-lang/rust/pull/150966)

You have to use `Arch::PowerPC64` and check the endian.

Gelbpunkt committed:
rustc_target: callconv: powerpc64: Use the ABI set in target options instead of guessing

All PowerPC64 targets except AIX explicitly set the ABI in the target
options. We can therefore stop hardcoding the ABI to be used based on
the target environment or OS, except for the AIX special case.

The fallback based on endianness is kept for the sake of compatibility
with custom targets.

This makes it so that big endian targets not explicitly accounted for
before (powerpc64-unknown-openbsd) and targets that don't use the
expected default ABI (ELFv2 Glibc targets) use the correct ABI in the
calling convention code.

Gelbpunkt committed:
rustc_target: spec: Ensure that a valid llvm_abiname value is set for PowerPC64(LE)

PowerPC64 ELF targets (effectively anything that isn't AIX) use either
the ELFv1 or ELFv2 ABI. The ELFv1 ABI is only specified for big endian
targets, while ELFv2 can be used by both little- and big-endian targets.
Make sure that, if an LLVM ABI is set, it is set to one of the two. AIX
does not set an LLVM ABI name, so ensure that AIX targets don't set
anything other than an empty ABI name.

Gelbpunkt commented:
> However, do we have to worry about this breaking the ABI for anyone already using these targets?

Previously, this hardcoded the list of big-endian targets using ELFv2 to musl and FreeBSD, which was not kept up to date (OpenBSD is missing from here). Since all PowerPC64 targets now specify their ABI, we can stop hardcoding the list. It will not break the ABI for any of them, AFAICT this is strictly a fix for the OpenBSD target and makes it so there's one less part of the code we need to worry about when changing or adding new targets.

jhpratt commented:
yielding to encompassing rollup

@bors yield
